pkg/instrumentation/sdk.go

```diff
@@ -431,48 +430,88 @@ func chooseServiceName(pod corev1.Pod, useLabelsForResourceAttributes bool, reso
 // The precedence is as follows:
 // 1. annotation with key resource.opentelemetry.io/<resource>.
 // 2. label with key labelKey.
-func chooseLabelOrAnnotation(pod corev1.Pod, useLabelsForResourceAttributes bool, resource attribute.Key, labelKey string) string {
+func chooseLabelOrAnnotation(pod corev1.Pod, useLabelsForResourceAttributes bool, resource attribute.Key, labelKeys []string) string {
 	if v := pod.GetAnnotations()[(constants.ResourceAttributeAnnotationPrefix + string(resource))]; v != "" {
 		return v
 	}
 	if useLabelsForResourceAttributes {
-		if v := pod.GetLabels()[labelKey]; v != "" {
-			return v
+		for _, labelKey := range labelKeys {
+			if v := pod.GetLabels()[labelKey]; v != "" {
+				return v
+			}
 		}
 	}
 	return ""
 }
 
 // chooseServiceVersion returns the service version to be used in the instrumentation.
-// The precedence is as follows:
-// 1. label or annotation with key "service.version" or "app.kubernetes.io/version".
-// 2. image tag (by splitting image string on ":" and extracting final element from resulting array).
+// See https://github.com/open-telemetry/semantic-conventions/blob/main/docs/non-normative/k8s-attributes.md#how-serviceversion-should-be-calculated
 func chooseServiceVersion(pod corev1.Pod, useLabelsForResourceAttributes bool, index int) string {
 	v := chooseLabelOrAnnotation(pod, useLabelsForResourceAttributes, semconv.ServiceVersionKey, constants.LabelAppVersion)
 	if v != "" {
 		return v
 	}
-	parts := strings.Split(pod.Spec.Containers[index].Image, ":")
-	tag := parts[len(parts)-1]
-	//guard statement to handle case where image name has a port number
-	if strings.Contains(tag, "/") {
+	var err error
+	v, err = parseServiceVersionFromImage(pod.Spec.Containers[index].Image)
+	if err != nil {
 		return ""
 	}
-	return tag
+	return v
+}
+
+// chooseServiceNamespace returns the service.namespace to be used in the instrumentation.
+// See https://github.com/open-telemetry/semantic-conventions/blob/main/docs/non-normative/k8s-attributes.md#how-servicenamespace-should-be-calculated
+func chooseServiceNamespace(pod corev1.Pod, useLabelsForResourceAttributes bool, namespaceName string) string {
+	namespace := chooseLabelOrAnnotation(pod, useLabelsForResourceAttributes, semconv.ServiceNamespaceKey, nil)
+	if namespace != "" {
+		return namespace
+	}
+	return namespaceName
+}
+
+var cannotRetrieveImage = errors.New("cannot retrieve image name")
+
+// parseServiceVersionFromImage parses the service version for differently-formatted image names
+// according to https://github.com/open-telemetry/semantic-conventions/blob/main/docs/non-normative/k8s-attributes.md#how-serviceversion-should-be-calculated
+func parseServiceVersionFromImage(image string) (string, error) {
```

Comment:
Could you add some unit tests for these new functions?

Reply:
it's here - and many tests:

image:

https://github.com/open-telemetry/opentelemetry-operator/blob/8d35a59c9f1ab26df4a6a6765cc2daac6763fbb8/pkg/instrumentation/sdk_test.go#L2645-L2683

namespace:

https://github.com/open-telemetry/opentelemetry-operator/blob/8d35a59c9f1ab26df4a6a6765cc2daac6763fbb8/pkg/instrumentation/podmutator_test.go#L853

Reply:
I don't see `service.version` checked anywhere in there. You've changed the behaviour for that attribute, so I'd like to see that reflected in a unit test. Plus. you've added a new function which doesn't look trivial to me, so it should have some basic tests to show that it actually does what the docstring claims.

Reply:
> I don't see `service.version` checked anywhere in there. You've changed the behaviour for that attribute, so I'd like to see that reflected in a unit test. 

here's one: https://github.com/open-telemetry/opentelemetry-operator/blob/8d35a59c9f1ab26df4a6a6765cc2daac6763fbb8/pkg/instrumentation/sdk_test.go#L229

It also shows in the diff when you open that file.

> Plus. you've added a new function which doesn't look trivial to me, so it should have some basic tests to show that it actually does what the docstring claims.

this is the test https://github.com/open-telemetry/opentelemetry-operator/blob/8d35a59c9f1ab26df4a6a6765cc2daac6763fbb8/pkg/instrumentation/sdk_test.go#L2645-L2683

or do you mean another function?

Reply:
Ah, I see it, my bad. Completely missed that file was being changed at all during review.